dbms/src/Interpreters/ThreadStatusExt.cpp

```diff
@@ -118,6 +129,74 @@ void ThreadStatus::finalizePerformanceCounters()
     }
 }
 
+namespace {
+    void queryProfilerTimerHandler(int /* sig */, siginfo_t * /* info */, void * context) {
+        DB::WriteBufferFromFileDescriptor out(TracePipe::instance().write_fd);
+
+        std::string queryID = CurrentThread::getCurrentQueryID();
+
+        DB::writePODBinary(*reinterpret_cast<const ucontext_t *>(context), out);
+        DB::writeStringBinary(queryID, out);
+        out.next();
+    }
+}
+
+void ThreadStatus::initQueryProfiler() {
+    if (!query_context) {
+        LOG_INFO(log, "Query profiler disabled - no context");
+        return;
+    }
+
+    const auto pause_signal = SIGALRM;
+
+    struct sigevent sev;
+    sev.sigev_notify = SIGEV_THREAD_ID;
+    sev.sigev_signo = pause_signal;
+    sev._sigev_un._tid = os_thread_id;
+    // TODO(laplab): get clock type from settings
+    if (timer_create(CLOCK_REALTIME, &sev, &query_profiler_timer_id)) {
+        throw Poco::Exception("Failed to create query profiler timer");
+    }
+
+    // TODO(laplab): get period from settings
+    struct timespec period{.tv_sec = 0, .tv_nsec = 200000000};
+    struct itimerspec timer_spec = {.it_interval = period, .it_value = period};
+    if (timer_settime(query_profiler_timer_id, 0, &timer_spec, nullptr)) {
+        throw Poco::Exception("Failed to set query profiler timer");
+    }
```

Comment:
Actually we want to be able to setup two timers simultaneously (CPU time and real time).